### PR TITLE
RMLNQ-107 Don't use Enum.IsDefined in EvaluatableTreeFindingExpressionVisitor.

### PR DESCRIPTION
**Pull request by anpete**
Identified as a hot-path during EF Core profiling. We move to a simple bounds check against the min and max values of ExpressionType.

### CODE REVIEW

**Comment by MichaelKetting** [comment]
Thanks for the PR! Review coming soon :)

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs [hunk 1]
```diff
@@ -61,6 +61,9 @@ public sealed class EvaluatableTreeFindingExpressionVisitor : RelinqExpressionVi
       return visitor._partialEvaluationInfo;
     }
 
+    // PERF: Avoids using Enum.IsDefined, which is really slow.
+    private static readonly ExpressionType s_maxExpressionType = Enum.GetValues(typeof (ExpressionType)).Cast<ExpressionType>().Max();
```
Comment (MichaelKetting) [inline Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: Please add a whitespace after Enum.GetValues. We always keep a space when calling methods with parameters.

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs [hunk 2]
```diff
@@ -541,7 +544,7 @@ protected override Expression VisitRelinqUnknownNonExtension (Expression express
     /// <see cref="ExpressionVisitor"/> has a dedicated Visit method for it; otherwise, <see langword="false"/>. 
     /// Note that <see cref="ExpressionType.Extension"/>-type expressions are considered 'not supported' and will also return <see langword="false"/>.
     /// </returns>
-    private bool IsCurrentExpressionEvaluatable (Expression expression)
+    private static bool IsCurrentExpressionEvaluatable (Expression expression)
```
Comment (MichaelKetting) [inline Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: Did you change the method to static because of performance reasons? I know there's a small difference in execution speed between calling instance methods and static methods but is it actually measurable?

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs [hunk 3]
```diff
@@ -552,8 +555,10 @@ private bool IsCurrentExpressionEvaluatable (Expression expression)
         return IsCurrentExpressionEvaluatable (reducedExpression);
       }
 
-      // Note: In .NET 4.0, all expressions supplied by System.Core have a dedicated Visit method provided by System.Linq.Parsing.ExpressionVisitor.
-      return Enum.IsDefined (typeof (ExpressionType), expression.NodeType);
+      // PERF: We don't use Enum.IsDefined because it is too slow. We can hard-code ExpressionType.Add
+      // because it will never change. 
+      return expression.NodeType >= ExpressionType.Add
```
Comment (MichaelKetting) [inline Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: Would it be better to place the ExpressionType.Add value into a static field similiar to the max-value and add the comment regarding the unchanging nature there. That way, the code becomes more symmetrical.

Also, please keep the previous note, too, since it explains a different detail.

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs [hunk 4]
```diff
@@ -625,7 +630,7 @@ private bool IsCurrentExpressionEvaluatable (Expression expression)
     }
 #endif
 
-    private bool IsQueryableExpression (Expression expression)
+    private static bool IsQueryableExpression (Expression expression)
```
Comment (MichaelKetting) [inline Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: Same as above regarding static. If there is an actual benefit to using static vs. instance, I think a separate dedicated PR would be better. In the code base, we rarely (in theory: never) use static unless the method needs to be static so if static has a measurable benefit for EF Core, there's probably more code in the call chain that needs to be changed to static.

File: UnitTests/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitorTest.cs [hunk 5]
```diff
@@ -75,6 +75,23 @@ public void SimpleExpression_IsEvaluatable ()
...
+    public void ExpressionTypeEnum_HasNoHoles ()
+    {
+      // We made a perf-optimization in RMLNQ-107 by avoiding calling Enum.IsDefined.
+      // This test verifies our assumption that the ExpressionType enum does not have holes.
+
+      var expected = 0;
+
+      foreach (var value in Enum.GetValues(typeof(ExpressionType)))
+      {
+        if (expected++ != (int)value)
```
Comment (MichaelKetting) [inline UnitTests/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitorTest.cs]: Please add a whitespace before between cast and value. Also, please add whitespaces in line 85 and 89.

File: UnitTests/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitorTest.cs [hunk 6]
```diff
@@ -75,6 +75,23 @@ public void SimpleExpression_IsEvaluatable ()
...
+        {
+          Assert.Fail("Unexpected hole found in ExpressionType enum!");
```
Comment (MichaelKetting) [inline UnitTests/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitorTest.cs]: The assertion: please separate the effect (incrementing "expected") from the condition. We prefer to keep statements limited to one effect.  Then, please change it to an Assert.That (value, Is.EqualTo(exected), "message'). It's the more common style in our tests :)

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
Thanks again. I've added a couple of formatting requests and a question regarding static. Logic looks good! :)

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs (hunk 2 above)
Comment (anpete) [reply Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: Yeah, it was flagged by R#. You're right that perf is likely neglible, but another thing to consider is that it documents that the method is stateless, which can help with maintainability. Happy to revert this, just let me know.

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs (hunk 4 above)
Comment (anpete) [reply Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: As above, happy to revert. Interested to understand why you avoid statics like this.

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs (hunk 3 above)
Comment (anpete) [reply Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: This is just to avoid declaring another static field. Happy to change it if you would prefer it to be symmetrical.

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs (hunk 2 above)
Comment (MichaelKetting) [reply Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: Okay, great. Yes, please revert them. I agree about the explicit statelessness, but it's one of those code style things from way back that's spread throughout our entire codebase so I prefer the consistency.

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs (hunk 3 above)
Comment (MichaelKetting) [reply Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: Thanks.

File: Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs (hunk 4 above)
Comment (MichaelKetting) [reply Core/Parsing/ExpressionVisitors/TreeEvaluation/EvaluatableTreeFindingExpressionVisitor.cs]: To follow up from above :)

Historically speaking (and it's still in use today), we've had the rule to put the static members before all the instnace members. This always included the non-public stuff. So, finding the static members down in the instance section feels odd when you're trained to look for this sort of thing.

The maintainability: Yes, I did consider this on occasion, too, when it was especially important and there could be one or two places that are static and documented as static. But for the most part, I address this by simply making my types imutable. That way, it doesn't matter if a method accesses an instance field or not. Maybe I just got lucky, but the maintainability issue has never come up for me when I'm reasoning about the code. 

Funny thing, actually, seeing them as instance members actually makes them easier to put into context for me, given how I know they're just used by the instance members of my code.

Does that make any sort of sense to an outside set of eyes?

**Comment by MichaelKetting** [comment]
@anpete  Sorry for the delay. I just now realized that Github didn't notify me about the already updated PR. It's merged now.
